### PR TITLE
REF: use sanitize_array in Index.__new__

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

i.e. share code paths with Series constructor.

This isn't Technically Correct until #49348

### CODE REVIEW

**Comment by jbrockmendel** [comment]
> pandas/core/indexes/base.py:216:8: R1720: Unnecessary "elif" after "raise", remove the leading "el" from "elif" (no-else-raise)

@MarcoGorelli are you the right person to complain to about this?  i hereby complain.

**Comment by MarcoGorelli** [comment]
yup, removing this one now

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -531,12 +502,24 @@ def __new__(
...
-                # with e.g. a list [1, 2, 3] casting to numeric is _not_ deprecated
-                subarr = _maybe_cast_data_without_dtype(subarr)
-                dtype = subarr.dtype
-            return Index(subarr, dtype=dtype, copy=copy, name=name)
+            if not isinstance(data, (list, tuple)):
+                # we allow set/frozenset, which Series/sanitize_array does not, so
+                #  cast to list here
+                data = list(data)
+            if len(data) == 0:
+                # unlike Series, we default to object dtype:
```
Comment (mroeschke) [inline pandas/core/indexes/base.py:510]: Would we want to align this in the future?

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -208,6 +206,22 @@
 _dtype_obj = np.dtype("object")
 
 
+def _wrapped_sanitize(cls, data, dtype: DtypeObj | None, copy: bool):
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:209]: the strict_ints keyword we could get rid of after #49609.  the rest of this wrapping is mostly about giving Index-specific exception messages, so is pretty benign

(actually, at the time i refactored this function out i was calling it 3ish times within `Index.__new__`.  now that its only called once, we could inline it)

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:510]: i have no strong opinion on this

File: pandas/core/indexes/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/indexes/base.py:209]: Inlining can happen in a follow up

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Pull request merged**